### ISSUE TITLE
Source information can't be obtained in some scenarios in VS 2017 Update 5

### ISSUE THREAD

**Issue by Shyam-Gupta** [nunit/nunit3-vs-adapter#423]
VS 2017 Update 5 extracts source information for discovered tests using Roslyn APIs. It first determines whether it can obtain this information for the test projects. If yes then it passes following tag in runsettings to the adapter:
<CollectSourceInformation>false</CollectSourceInformation>
NUnit adapter 3.9.0 already honors this setting. 

However, for some scenarios, VS is not able to obtain this information:

Examples :
Case #1

    public class MyDataClass
    {
        public static IEnumerable<TestCaseData> TestCases
        {
            get
            {
                yield return new TestCaseData(12, 3).SetName("NewTestName").Returns(4);
                yield return new TestCaseData(12, 4).Returns(3);
            }
        }
    }
    [TestFixture]
    public class MyTests2
    {
        [Test, TestCaseSource(typeof(MyDataClass), "TestCases")]
        public int DivideTest(int n, int d)
        {
            return n / d;
        }
    }
 
Case #2

        [TestCase(TestName="foobar")]
        public void TestMethod()
        {
        }

**Reason:** In above mentioned cases, the fully qualified name for the test methods as reported by nunit doesn't correspond to a test case present in the source code. Hence using Roslyn APIs SourceInformationProvider is not able to find the test method and hence fails to fetch its source information.

Nunit adapter\engine which sends test method Fully Qualified Name, should be fixed to send correct test method name in above scenarios

**Comment by CharliePoole** [comment]
@Shyam-Gupta I updated your description to use backticks so that the XML is properly displayed in GitHub.

**Comment by CharliePoole** [comment]
One way to fix this if for NUnit to "fake" the FQN as suggested. Is there also the option for us to __provide__ the source information as part of the `TestCase` as we currently do? Would that be a better solution?

Our current code is providing source code information. Unless something has changed, it should be provided in the example cases cited. Can you indicate what is currently provided by the adapter in those cases?

**Comment by peterwald** [comment]
@CharliePoole providing source information from the adapter would certainly work, although that negates the performance improvement that this change was created to provide. 

Another option would be for the adapter to provide the Type & Method name in a well-known property on the test case (in the property bag). Then on the VS side, we can use the Type & Method to do the source information lookup without requiring the adapter to load symbols.

**Comment by CharliePoole** [comment]
@peterwald Ironically, I've proposed this to one of the teams in the past. We maintained class and method name separately in our XML so that we could figure out the source location. We have never used our test names for that purpose. If this is a practical option, just tell us where to put them!!! I'll even help although I'm not actually working on the project any longer!

**Comment by CharliePoole** [comment]
Also important to remember: whatever change we make here, the explorer as well as vstest should continue to use location information where it is provided by the adapter. Otherwise, it will break people using an older version of the adapter with a newer VS.

**Comment by peterwald** [comment]
Agreed. In addition to the Type and Method, we need to keep track of the method signature as well to distinguish between overloaded methods.

I believe the information we will need is:
TypeName
MethodName (with backtick notation if generic)
List of parameter TypeNames

**Comment by CharliePoole** [comment]
That's all available in the framework but not necessarily in the current XML that is returned to the adapter defining each test.

TypeName - is present
MethodName - the raw name is there, I believe without the backtick notation for generics.
Parameter Types are not provided in the XML at this time although we have discussed doing it.

@rprouse @OsirisTerje Would you favor providing more info from the framework, either sufficient to calculate source location or source location itself?

@peterwald Not sure why you feel NUnit providing the source location info would result in inefficiency. Could you expand that idea?

**Comment by OsirisTerje** [comment]
@CharliePoole I'm in favour of adding the parameters type information.  It would not slow down discovery with the new Roslyn discovery anyway.  It would give a small, but I believe insignificant increase in time during the execution.

**Comment by peterwald** [comment]
@CharliePoole RE: inefficiency.... You have to parse the symbol files in order to get source location info, which is not particularly fast. That is the reason we have the `<CollectSourceInformation>false</CollectSourceInformation>` runsetting.

**Comment by CharliePoole** [comment]
Ah. I was assuming it should be an option wherever it was done. I can imagine other runners that want to have source information.

**Comment by rprouse** [comment]
I am fine with updating the framework accordingly, but won't this cause problems for users who have not updated the framework forcing us to not pass the required info? @peterwald can you handle that?

Also, do you need the generics info?

**Comment by Shyam-Gupta** [comment]
@rprouse We can handle that in Test Window code. If the required info is not present, then we can fall back to current way of obtaining source information, i.e., through FQN.

Yes we will need generics info. Count of generic parameters is required to select the right method.

**Comment by sliddington** [comment]
Apologies if it is inappropriate to ask this here.

Is there a workaround to this issue? - the majority of our tests are named this way. Can we tell VS to do things the old slow way for everything until this is fixed? Obviously one way is to remove all the test names but then some test cases will clash I think through objects in parameters being indistinguishable and appearing to VS as the same test.

Also I think it would be warranted to add this to https://github.com/nunit/docs/wiki/Known-Problems

**Comment by rprouse** [comment]
@sliddington, does turning off source based discovery in **Tools | Options | Test** work for you?

![image](https://user-images.githubusercontent.com/493828/37571000-76a477b8-2acd-11e8-9d39-12db5c424894.png)

**Comment by sliddington** [comment]
With that setting off I still get the "Unable to fetch source Information for test method" warnings in the test output pane during test discovery. I now have the problem that the test explorer window does not seem to finish test discovery - although the tests output says it has. This means I can't even try running tests to see if it would work.

If I restart VS the test discovery does the same again, if I turn the setting back on and restart it gradually fills up the test list in test explorer and I am able to run tests but can't run the named ones.
